src/test/scala/rise/autotuning.scala

```diff
@@ -0,0 +1,594 @@
+                              let(toLocal(makePair(
+                                p14._1 |> join |> split(v6) |> // ((v8 x v5) /^ v6).v6.f
+                                  mapLocal(1)(asScalar o mapLocal(0)(id) o asVectorAligned(4)) |>
+                                  join |> split(v5)
+                              )( // v8.v5.f
+                                p14._2 |> // v8.v7.f
+                                  mapLocal(1)(asScalar o mapLocal(0)(id) o asVectorAligned(4))
+                              )))
+                                be (p15 =>
+                                zip(p13)(split(v4)(transpose(p15._1))) |> // (v5/^v4).((v7/^v3).v4.v3.f x v4.v8.f)
+                                  mapLocal(1)(fun(p16 =>
+                                    zip(p16._1)(split(v3)(transpose(p15._2))) |> // (v7/^v3).(v4.v3.f x v3.v8.f)
+                                      mapLocal(0)(fun(p17 =>
+                                        zip(transpose(p16._2))(transpose(p17._2)) |> // v8.(v4.f x v3.f)
+                                          oclReduceSeq(AddressSpace.Private)(fun((p19, p20) =>
+                                            // v4.v3.f x (v4.f x v3.f)
+                                            let(toPrivate(makePair(mapSeq(id)(p20._1))(mapSeq(id)(p20._2))))
+                                              be (p21 =>
+                                              zip(p19)(p21._1) |> // v4.(v3.f x f)
+                                                mapSeq(fun(p22 =>
+                                                  zip(p22._1)(p21._2) |> // v3.(f x f)
+                                                    mapSeq(fun(p23 =>
+                                                      p23._1 + (p22._2 * p23._2)
+                                                    ))
+                                                ))
+                                              )
+                                          ))(p17._1 // v4.v3.f
+                                            |> mapSeq(mapSeq(id)) // TODO: think about that
+                                          ) |> mapSeq(mapSeq(id)) // TODO: think about that
+                                      ))
+                                  ))
+                                )
+                            ))(
+                              generate(fun(_ =>
+                                generate(fun(_ =>
+                                  generate(fun(_ =>
+                                    generate(fun(_ => lf32(0.0f))))))))) |>
+                                mapLocal(1)(mapLocal(0)(mapSeq(mapSeq(id))))
+                            ) |> // (v5/^v4).(v7/^v3).v4.v3.f
+                            mapLocal(1)(mapLocal(0)(mapSeq(asScalar o mapSeq(id) o asVector(4)))) |>
+                            map(transpose) |> join |> map(join) |> transpose // v7.v5.f
+                        )) |> join |> transpose // v5.M.f
+                    )) |> join // N.M.f
+                ))))))))
```

Comment:
it feels like this same kernel is duplicated at least three times in the code, can we factorize this?

Reply:
Yes, we can. But we use this kernel here only to test the constraints collection and constraints checking for the autotuning. In [mm.scala](https://github.com/rise-lang/shine/blob/autotuning_merge/src/main/scala/apps/mm.scala) we have the non-autotuning version.

Reply:
You could do something like:
```scala
def mmNVIDIAWithParams(v3: Nat, ...): ToBeTyped[Expr] = ...

val mmKernel = tuningParam("v3", (v3: Nat) => ... mmNVIDIAWithParams(v3, ...))

val mmNVIDIA = mmNVIDIAWithParams(v3 = 4, ...)
```

Reply:
You could also do:
```
val mmNVIDIA = mmKernel.subsituteNats(Map(v3 -> 4, ...))
```

Reply:
I created a factorized version. Looks much better having this duplicated 60 lines deleted.